zebra-network/src/peer/connection.rs

```diff
@@ -127,34 +146,69 @@ impl Handler {
                     }
                 };
                 if items.iter().all(hash_in_hashes) {
-                    Handler::Finished(Err(PeerError::NotFound(items)))
+                    // all hashes have a transaction or a `NotFound` entry
+                    if !transactions.is_empty() {
+                        // TODO: does the caller need a list of missing transactions?
+                        Handler::Finished(Ok(Response::Transactions(transactions)))
+                    } else {
+                        Handler::Finished(Err(PeerError::NotFound(items)))
+                    }
                 } else {
+                    // waiting for more transactions or a complete `NotFound` message
+                    if items.iter().any(hash_in_hashes) {
+                        debug!(
+                            not_found_transactions = ?items.iter().cloned().filter(hash_in_hashes).count(),
+                            remaining_transactions = ?hashes.len(),
+                            "notfound containing some of the remaining requested transactions");
+                    }
                     ignored_msg = Some(Message::NotFound(items));
                     Handler::TransactionsByHash {
                         hashes,
                         transactions,
                     }
                 }
             }
+            // `zcashd` returns blocks sequentially (but not necessarily immediately).
+            // It silently skips missing blocks, rather than using `NotFound`:
+            // https://github.com/zcash/zcash/blob/e7b425298f6d9a54810cb7183f00be547e4d9415/src/main.cpp#L5523
             (
                 Handler::BlocksByHash {
                     mut hashes,
                     mut blocks,
                 },
                 Message::Block(block),
             ) => {
+                // assumptions:
+                //   - the block list is contiguous
+                //   - missing blocks are silently skipped (rather than using `NotFound`)
                 if hashes.remove(&block.hash()) {
+                    // we are in the middle of the contiguous block response
                     blocks.push(block);
+                    if hashes.is_empty() {
+                        Handler::Finished(Ok(Response::Blocks(blocks)))
+                    } else {
+                        Handler::BlocksByHash { hashes, blocks }
+                    }
                 } else {
+                    // we are before or after the contiguous block response
                     ignored_msg = Some(Message::Block(block));
-                }
-                if hashes.is_empty() {
-                    Handler::Finished(Ok(Response::Blocks(blocks)))
-                } else {
-                    Handler::BlocksByHash { hashes, blocks }
+                    if !blocks.is_empty() {
```

Comment:
We should unconditionally end the request if we get any block that isn't in our request. Terminating a request early is better than hanging on a request that will never be answered.
```suggestion
```

Reply:
If the peer answers our request later, the items we asked for will end up being processed by the inbound service - so we won't lose any items.

Reply:
Resolved by 221e1f87e5c827a0f8575ca569f4c5e0aa6aa5d0